### PR TITLE
add license mit

### PR DESCRIPTION
**Pull request by oxarbitrage**
<del>- [ ] Need to add licensing info for include/fc/stacktrace.hpp.</del>

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2017 bitshares-fc
```
Comment (pmconrad) [inline LICENSE]: That doesn't make sense, bitshares-fc is not an entity that can own copyright.

AFAIK the only use-case for the copyright owner is when it comes to defending the license in court. Which is unlikely to happen, given the liberal nature of the MIT license as such.

We could use "The BitShare Core Development Team" (although that's not really a legally existing entity either), or perhaps the BBF. @xeroc  @ryanRfox opinions?

**Comment by ryanRfox** [comment]
I'm not sure the proper entity to assign the copyright to either. I agree "bitshares-fc" does not seem the proper entity. bitshares-core introduces a stub for CONTRIBUTORS (yet remains blank). Looking at the fork hierarchy, none of the FC repositories contain a LICENSE file. My intent will always be to honor all prior copyright holders and aim to offer all code committed to the BitShares organization under MIT license (bitsharaes-core, bitshares-ui, bitshares-fc, etc.). I am keen to hear from @xeroc what the BBF may recommend based on their ongoing legal work toward rewriting the whitepaper(s).

**Comment by pmconrad** [comment]
Regarding history - the library is public domain, according to this post from BM: https://bitsharestalk.org/index.php?topic=8571.msg142379#msg142379
> Stuff we write has no restrictions except those imposed on it by 3rd party dependencies. 

Regarding 3rd party dependencies - please do not merge this PR before libreadline has been replaced.

**Comment by xeroc** [comment]
Given the nature of the MIT license, does it actually matter who is written next to "copyright"?

**Comment by abitmore** [comment]
> Given the nature of the MIT license, does it actually matter who is written next to "copyright"?

Sure. The MIT license says:
>The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

**Comment by abitmore** [comment]
FYI Steem just moved fc into the main repo: https://github.com/steemit/steem/pull/2145#event-1485845034. EOS has already done so long before. Perhaps we can do the same.

**Comment by abitmore** [comment]
>Regarding history - the library is public domain, according to this post from BM: https://bitsharestalk.org/index.php?topic=8571.msg142379#msg142379
>
>>Stuff we write has no restrictions except those imposed on it by 3rd party dependencies.

IANAL, strictly speaking, I don't think BM's post was appropriate, nor implicitly included fc.

1. the post was posted on `2014-10-31`, at that time, according to the Bitshares Social Consensus associated with AGS, AGS "holders" and PTS holders would receive at least 10% share-drop if the code is used to launch a new chain. That said, there may be no restriction about how to distribute the code or binary built from it, but there was at least a restriction about how to use it. Licensing the software under public domain was not of the AGS holders' interest, although strictly speaking BM has the right to do so since "AGS is donation with no strings attached".

2. in regards to fc, in the post BM used a word "we" may be referring to the team, and a word "write" which was present tense, both are ambiguous. Latest commit of BM's fc library is on `Aug 19, 2013`, which is clearly a product of his own but not the team. When he was saying "we write", it's not clear whether his own product (fc) was implicitly included, since it (fc) may be one of the "3rd party dependencies".

Of course, we can argue that since fc used readline so fc should be licensed under AGPL. However, that means the product using fc (e.g. bitshares-core) should also be licensed under AGPL, which is perhaps not what we want.

Personally I still prefer that BM clearly license the bm/fc repository under MIT, it's just a few clicks if he really wants to do so.

**Comment by pmconrad** [comment]
IMO the social consensus is irrelevant here, since it never was a legally binding agreement.

The AGPL code was removed some time after that thread. The only remaining "problem" is the readline dependency. Yes, the transitive nature of GPL means that fc and in turn bitshares-core (as well as all the other graphene chains) have to be licensed under the GPL as well. I think we do not want to do that, that's why we should replace readline.

The rest of the fc code either has its own license (BSD or MIT mostly), or was created by BM/Invictus/Cryptonomex. The latter part is what we must be concerned about.

> Steem just moved fc into the main repo. EOS has already done so long before.

This opens another argument: If the code has been moved into the main repo, it is now also covered by the license of that repo. Not sure about Steem, but BM's current involvement in EOS means he probably has the authority to make that move. EOS is MIT-licensed as well, right? We can then either try to move to the current EOS version of fc and use the EOS license, or we take that move as an indication that a wide interpretation of BM's statement from 2014 is in order and that our fc version is indeed public domain.

**Comment by abitmore** [comment]
According to https://github.com/EOSIO/eos/blob/master/LICENSE.txt, EOS will be licensed under MIT after June 1, 2018, aka the scheduled launch date, but not before. By the way, the change was made on Jun 21, 2017, before that date, it's MIT.

**Comment by abitmore** [comment]
Steem is now MIT licensed. However, there was a drama when changing to MIT, which I guess is related to FC. It's clearly written in https://github.com/steemit/steem/blob/master/LICENSE.md:

>The following license applies to code contained within this repository that is created by Steemit, Inc. Other copy right holders have licensed dependencies such as Graphene, FC, and Boost under their own individual licenses.
>
>The MIT License
>...

So even if the main repo is MIT licensed, FC is not included.

**Comment by pmconrad** [comment]
Hm, so steem did their due diligence...

There's a significant delta between the pre-2017-06 version of EOS-fc and ours. In particular, the threading stuff has been ripped out. Using this is not an option.

We should merge #14 once @jmjatlanta has pushed his latest changes, then contact BM and
1. offer #14 so he can get rid of the GPL problem he has with eos-fc, and
2. ask him to confirm that we can re-license bitshares-fc as MIT.

**Comment by pmconrad** [comment]
Excellent!

**Comment by abitmore** [comment]
https://github.com/bitshares/bitshares-fc/pull/14#issuecomment-368663557 quoted here:
> Note: The new editline library has a liberal license, but still requires documentation changes. See the license at https://github.com/troglobit/editline/blob/master/LICENSE

**Comment by abitmore** [comment]
Thank you @pmconrad. Now we only need to finalize the main license file.

File: LICENSE (hunk 1 above)
Comment (clockworkgr) [reply LICENSE]: Don't think it has to be a legal entity at all.

"BitShares Core Development Team" or "BitShares Developers" would be fine I think.

Comment (abitmore) [reply LICENSE]: I suggest "Copyright (c) 2013-2018 Daniel Larimer and contributors."

**Comment by abitmore** [comment]
Specifically, in the main license file, I think we need to mention the files in the licenses directory.

**Comment by abitmore** [comment]
About the [Bloom Filter Library](https://github.com/bitshares/bitshares-fc/blob/master/include/fc/bloom_filter.hpp) included in this repository:

```
/*
 *********************************************************************
 *                                                                   *
 *                           Open Bloom Filter                       *
 *                                                                   *
 * Author: Arash Partow - 2000                                       *
 * URL: http://www.partow.net                                        *
 * URL: http://www.partow.net/programming/hashfunctions/index.html   *
 *                                                                   *
 * Copyright notice:                                                 *
 * Free use of the Open Bloom Filter Library is permitted under the  *
 * guidelines and in accordance with the most current version of the *
 * Common Public License.                                            *
 * http://www.opensource.org/licenses/cpl1.0.php                     *
 *                                                                   *
 *********************************************************************
*/
```

**Comment by clockworkgr** [comment]
Original FC library MIT licensed -> https://github.com/bytemaster/fc/blob/phoenix/LICENSE.md

**Comment by abitmore** [comment]
Need to add licensing info for include/fc/stacktrace.hpp . OP updated.

**Comment by pmconrad** [comment]
Did some more research.
* The license in BM's repo is more or less useless to us:
    * Copyright says 2010, but earliest commit is 2012
    * Only a small number of files in BM's repo is identical to ours
* The license in Steem is also useless, it explicitly does not cover fc.
* The license in EOS repo covers more of our files, BUT I'm very sure that the original authors (including myself) were not asked to re-license their code.

Our history crosses several organizations: BitShares -> Steem -> Cryptonomex -> Bytemaster, and somewhere in between there was also Invictus Innovations who created BitShares-1.

IMO it is close to impossible to straighten this out. Our best bet might be to
* Add a license file that covers our own code plus all future contributions
* Get rid of as much fc code as possible (see https://github.com/bitshares/bitshares-core/issues/1584)
* Replace code that we can't remove with our own improvements under the given license

**Comment by pmconrad** [comment]
* Rebased
* merged LICENSE and README.LICENSE into LICENSE.md
* fixed the stacktrace license header because it bears no resemblance to either the original nor the steem version, since it was essentially redesigned around boost::stacktrace by @jmjatlanta

**Comment by abitmore** [comment]
Thank you @pmconrad for the efforts. I'd suggest that we move the changes about bloom filter and cmake files to new pull requests, only leave license related changes in this pull request.

**Comment by pmconrad** [comment]
Done and rebased.

**Comment by abitmore** [comment]
Looks good to me. Thanks!

**Comment by pmconrad** [comment]
Other opinions? @jmjatlanta @xeroc @oxarbitrage ?

File: LICENSE.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+The following license only applies to code that was contributed to this repository under the regime of BitShares worker proposals and/or BitShares delegate workers. Other code may be licensed indirectly by our "sister" repositories https://github.com/EOSIO/fc and https://github.com/bytemaster/fc/tree/phoenix .
```
Comment (oxarbitrage) [inline LICENSE.md]: Can we replace the word regime ?

Comment (pmconrad) [reply LICENSE.md]: done

**Comment by oxarbitrage** [comment]
Looks good to me but i don't have permissions to approve.

**Comment by oxarbitrage** [comment]
> Looks good to me but i don't have permissions to approve.

The original PR was opened by me on Feb 2018. @pmconrad used the same PR and modified recently. That is why i cant approve, not a permission issue.

Sorry for the misunderstanding.

**Review by abitmore — APPROVED** [review]

**Pull request merged**